### PR TITLE
 Feature - TriNguyen - CustomView

### PR DESCRIPTION
**Pull request by mvn-tringuyen-dn**

### CODE REVIEW

File: 05_CustomView/CustomView/CustomView/Ex1/MyAvatarView.swift [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+//
+//  Created by tri.nguyen on 04/04/2022.
+//
+
+import UIKit
+
+// MARK: - Protocol
+
+
+class MyAvatarView: UIView {
```
Comment (kiyochinh) [inline 05_CustomView/CustomView/CustomView/Ex1/MyAvatarView.swift:13]: `final class`

File: 05_CustomView/CustomView/CustomView/Ex1/MyAvatarView.swift [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    }()
+    
+    private lazy var myButton: UIButton = {
+        let button = UIButton()
+        button.addTarget(self, action: #selector(tapButtonTouchUpInside), for: .touchUpInside)
+        return button
+    }()
+    
+    // Objc
+    @objc func tapButtonTouchUpInside() {
```
Comment (kiyochinh) [inline 05_CustomView/CustomView/CustomView/Ex1/MyAvatarView.swift:38]: `@objc private func`

File: 05_CustomView/CustomView/CustomView/HomeViewController.swift [hunk 3]
```diff
@@ -0,0 +1,51 @@
+//
+//  HomeViewController.swift
+//  CustomView
+//
+//  Created by tri.nguyen on 04/04/2022.
+//
+
+import UIKit
+
+class HomeViewController: UIViewController {
```
Comment (kiyochinh) [inline 05_CustomView/CustomView/CustomView/HomeViewController.swift:10]: `final`

File: 05_CustomView/CustomView/CustomView/HomeViewController.swift [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+    // MARK: - Life cycle
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        setupUI()
+    }
+}
+
+// MARK: - Private Functions
```
Comment (kiyochinh) [inline 05_CustomView/CustomView/CustomView/HomeViewController.swift:19]: Xuống 1 dòng cho đẹp
```
// MARK: -

```

**Review by kiyochinh — COMMENTED** [review]
_

File: 05_CustomView/TriNC/CustomView/CustomView/Ex1/MyAvatar/MyAvatarView.swift [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+//  CustomView
+//
+//  Created by tri.nguyen on 07/04/2022.
+//
+
+import Foundation
+import UIKit
+
+// 2: Tạo Protocol MyDelegateAvatar
+protocol MyAvatarDelegate {
```
Comment (mvn-tanhuynh-dn) [inline 05_CustomView/TriNC/CustomView/CustomView/Ex1/MyAvatar/MyAvatarView.swift:12]: Rename to MyAvatarViewDelegate

File: 05_CustomView/TriNC/CustomView/CustomView/Ex1/MyAvatar/MyAvatarView.swift [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    func didTap(didTap with: MyAvatarView, countUser: Int)
```
Comment (mvn-tanhuynh-dn) [inline 05_CustomView/TriNC/CustomView/CustomView/Ex1/MyAvatar/MyAvatarView.swift:13]: Create Action enum in MyAvatarView and rename function to format:
func view(_ view: MyAvatarView, needsPerform action: MyAvatarView.Action)

File: 05_CustomView/TriNC/CustomView/CustomView/Ex2/MySlider/MySliderView.swift [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+//  CustomView
+//
+//  Created by tri.nguyen on 11/04/2022.
+//
+
+import Foundation
+import UIKit
+
+// MARK: - Protocol
+protocol MySliderDelegate: class {
```
Comment (mvn-tanhuynh-dn) [inline 05_CustomView/TriNC/CustomView/CustomView/Ex2/MySlider/MySliderView.swift:12]: Rename to MySliderViewDelegate

File: 05_CustomView/TriNC/CustomView/CustomView/Ex2/MySlider/MySliderView.swift [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+        let yBlueView = (CGFloat(100 - value) * (containerView.frame.maxY - 25) / 100)
+        let heightView = CGFloat(value) * containerView.frame.size.height / 100
+        blueView.frame = CGRect(x: blueView.frame.origin.x, y: yBlueView, width: blueView.frame.size.width, height: heightView)
+        thumbView.center.y = yBlueView
+        let percent = (blueView.frame.size.height / containerView.frame.size.height) * 100
+        percentLabel.text = "\(percent)%"
+    }
+    
+    // MARK: - Touch Moved
+    override func touchesMoved(_ touches: Set<UITouch>, with event: UIEvent?) {
```
Comment (mvn-tanhuynh-dn) [inline 05_CustomView/TriNC/CustomView/CustomView/Ex2/MySlider/MySliderView.swift:48]: override should call super.touchesMoved method

File: 05_CustomView/TriNC/CustomView/CustomView/Ex4/DatePickerView/DatePickerView.swift [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+//
+//  Created by tri.nguyen on 12/04/2022.
+//
+
+import Foundation
+import UIKit
+
+// MARK: - Protocol
+protocol DatePickerViewDelegate: class {
+    func getDate(_ view: DatePickerView, date: String)
```
Comment (mvn-tanhuynh-dn) [inline 05_CustomView/TriNC/CustomView/CustomView/Ex4/DatePickerView/DatePickerView.swift:13]: Please create enum Action and follow to convention:
func view(_ view: DatePickerView, needPerform action: DatePickerView.Action)

File: 05_CustomView/TriNC/CustomView/CustomView/HomeViewController.swift [hunk 10]
```diff
@@ -0,0 +1,50 @@
+//
+//  HomeViewController.swift
+//  CustomView
+//
+//  Created by tri.nguyen on 04/04/2022.
+//
+
+import UIKit
+
+class HomeViewController: UIViewController {
```
Comment (mvn-tanhuynh-dn) [inline 05_CustomView/TriNC/CustomView/CustomView/HomeViewController.swift:10]: Add final keyword

**Review by mvn-tanhuynh-dn — APPROVED** [review]
LGTM!

**Pull request merged**